### PR TITLE
Add a UI to show ingestion events

### PR DESCRIPTION
**Pull request by philmcmahon**
## What does this change?
Following https://github.com/guardian/giant/pull/117, this PR adds an API endpoint and associated client side code to display a dashboard of ingestion events.

Currently there are two dashboards - 'my uploads' and 'all ingestion events', the latter of which can only be seen by admins. The former allows a user to explore uploads they have made to workspaces, whilst the latter is intended for admins to be able to explore all the ingestion events that have occurred to e.g. help debug a user who has reported a particular issue.

However, pending some work to auto-delete ingestion events after a period of time, at present the permissions are more locked down such that you can only see an ingestion dashboard if you already have permissions to view a particular collection. Admins will have to add themselves to particular collection before trying to debug issues with it.

The most complex part of this work is the new getEvents query in PostgresClient.scala. This was initially written by @zekehuntergreen - with the main complexity trying to get a one-blob-per-row view with all the relevant status updates for the different extractors run on a blob, taking account of the fact that a file can be ingested multiple times.

The frontend makes use of [Elastic UI](https://elastic.github.io/eui/) - there's probably an open question as to how heavily we want to start relying on EUI in giant due to the licensing concerns, but this was a quick way to get a nice table going and, as it's buried in the settings, I'm not too concerned about it looking a big jarring compared to the rest of giant's UI. An alternative  would be to use something like https://chakra-ui.com/getting-started/with-react-table

Right now I think the UI still needs a lot of work but has some basic useful elements. The aim of this PR is to provide a building block from which individual pieces of work can be done. For example, there's currently no rendering of errrors in the UI, which feels like it might be useful.

Missing UI features:
 - Paging
 - Search
 - Store current dashboard view in UI so refresh doesn't reset everything (right now refresh is the only way to get updates)
 - Much more - we should do an ingestion and work out what would be useful to see
 - auto update (particularly annoying given that when you refresh you lose the page you were on)

It looks like this:
Screenshot 2023-08-22 at 11 33 42
Screenshot 2023-08-22 at 11 33 13

## How to test
Tested on playground, seems to work ok - obviously with lots of missing data

## Have we considered potential risks?
As these dashboards are fairly hidden away the main concern is security - which is discussed above

### CODE REVIEW

File: backend/app/commands/IngestFile.scala [hunk 1]
```diff
@@ -31,11 +31,14 @@ class IngestFile(collectionUri: Uri, ingestionUri: Uri, uploadId: String, worksp
       ingestion <- getIngestion()
 
       fileUri = FingerprintServices.createFingerprintFromFile(temporaryFilePath.toFile)
+      workspaceEvent = workspace.map(w =>
+        IngestionEvent.workspaceUploadEvent(fileUri, ingestionUri.value, w.workspaceName, EventStatus.Started))
+      _ = workspaceEvent.foreach(ingestionServices.recordIngestionEvent)
```
Comment (philmcmahon) [inline backend/app/commands/IngestFile.scala:36]: document this

File: frontend/src/js/components/IngestionEvents/IngestionEvents.tsx [hunk 2]
```diff
@@ -0,0 +1,221 @@
...
+    {collectionId, ingestId, workspaces, breakdownByWorkspace}: {
+        collectionId: string,
+        ingestId?: string,
+        workspaces: WorkspaceMetadata[],
+        breakdownByWorkspace: boolean
+    }) {
+    const [blobs, updateBlobs] = useState<BlobStatus[] | undefined>(undefined)
+
+    const ingestIdSuffix = ingestId && ingestId !== "all" ? `/${ingestId}` : ""
+    const [tableData, setTableData] = useState<IngestionTable[]>([])
```
Comment (philmcmahon) [inline frontend/src/js/components/IngestionEvents/IngestionEvents.tsx:161]: maybe put state together

File: frontend/src/js/components/IngestionEvents/IngestionEvents.tsx [hunk 3]
```diff
@@ -0,0 +1,221 @@
...
+}
+
+function IngestionEvents(
+    {collectionId, ingestId, workspaces, breakdownByWorkspace}: {
+        collectionId: string,
+        ingestId?: string,
+        workspaces: WorkspaceMetadata[],
+        breakdownByWorkspace: boolean
+    }) {
+    const [blobs, updateBlobs] = useState<BlobStatus[] | undefined>(undefined)
```
Comment (philmcmahon) [inline frontend/src/js/components/IngestionEvents/IngestionEvents.tsx:158]: explicit | undefined isn't necessary

File: frontend/src/js/components/IngestionEvents/IngestionEvents.tsx [hunk 4]
```diff
@@ -0,0 +1,221 @@
...
+    return {
+        workspacesMetadata: state.workspaces.workspacesMetadata,
+        currentUser: state.auth.token?.user,
+        collections: state.collections
+    };
+}
+
+function mapDispatchToProps(dispatch: GiantDispatch) {
+    return {
+        getCollections: bindActionCreators(getCollections, dispatch),
```
Comment (philmcmahon) [inline frontend/src/js/components/IngestionEvents/IngestionEvents.tsx:217]: maybe document why we need this

File: frontend/src/js/components/IngestionEvents/IngestionEvents.tsx [hunk 5]
```diff
@@ -0,0 +1,221 @@
...
+        </>
+}
+
+
+function mapStateToProps(state: GiantState) {
+    return {
+        workspacesMetadata: state.workspaces.workspacesMetadata,
+        currentUser: state.auth.token?.user,
+        collections: state.collections
+    };
```
Comment (philmcmahon) [inline frontend/src/js/components/IngestionEvents/IngestionEvents.tsx:212]: this is all wrong!

**Review by philmcmahon — COMMENTED** [review]
Comments from review session with @marjisound

File: frontend/src/js/components/IngestionEvents/IngestionEvents.tsx (hunk 3 above)
Comment (philmcmahon) [reply frontend/src/js/components/IngestionEvents/IngestionEvents.tsx:158]: I tried removing | undefined but it resulted in red underlining in intellij

File: backend/app/services/observability/Models.scala [hunk 6]
```diff
@@ -73,31 +75,79 @@ object EventDetails {
...
 
-case class BlobMetaData(ingestId: String, blobId: String, path: String, fileSize: Long)
 
-object BlobMetaData {
-  implicit val blobMetaDataFormat = Json.format[BlobMetaData]
+case class ExtractorStatus(extractorType: ExtractorType, statusUpdates: List[ExtractorStatusUpdate])
+object ExtractorStatus {
+  implicit val format = Json.format[ExtractorStatus]
+
+  def parseDbStatusEvents(extractors: Array[String], extractorEventTimes: Array[String], extractorStatuses: Array[String]): List[ExtractorStatus] = {
```
[the remark this answers is not in the document]
Comment (philmcmahon) [reply backend/app/services/observability/Models.scala:124]: I have written a single test. It took ages. I hate tests!

**Review by marjisound — APPROVED** [review]
All looks good, wonderful job 🎉 👏

**Pull request merged**